frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	}
+
+	/// Decode the length of the signed submissions without actually reading the entire struct into
+	/// memory.
+	///
+	/// Note that if you hold an instance of `SignedSubmissions`, this function does _not_
+	/// track its current length. This only decodes what is currently stored in memory.
+	pub fn decode_len() -> Option<usize> {
+		SignedSubmissionIndices::<T>::decode_len()
+	}
+
+	/// Insert a new signed submission into the set.
+	///
+	/// Returns `(inserted, removed)`. `inserted` is true when the submission was inserted.
+	/// `removed` is the removed weakest submission, if any.
+	///
+	/// In the event that the new submission is not better than the current weakest according
+	/// to `is_score_better`, we do not change anything.
+	pub fn insert(
+		&mut self,
+		submission: SignedSubmissionOf<T>,
+	) -> (bool, Option<SignedSubmissionOf<T>>) {
+		let weakest = match self.indices.try_insert(submission.solution.score, self.next_idx) {
+			Ok(Some(prev_idx)) => {
+				// a submission of equal score was already present in the set;
+				// no point editing the actual backing map as we know that the newer solution can't
+				// be better than the old. However, we do need to put the old value back.
+				self.indices
+					.try_insert(submission.solution.score, prev_idx)
+					.expect("didn't change the map size; qed");
+				return (false, None);
+			}
+			Ok(None) => {
+				// successfully inserted into the set; no need to take out weakest member
+				None
+			}
+			Err((score, insert_idx)) => {
+				// could not insert into the set because it is full.
+				// note that we short-circuit return here in case the iteration produces `None`.
+				// If there wasn't a weakest entry to remove, then there must be a capacity of 0,
+				// which means that we can't meaningfully proceed.
+				let (weakest_score, weakest_idx) = match self.indices.iter().next() {
+					None => return (false, None),
+					Some((score, idx)) => (*score, *idx),
+				};
+				let threshold = T::SolutionImprovementThreshold::get();
+
+				// if we haven't improved on the weakest score, don't change anything.
+				if !is_score_better(score, weakest_score, threshold) {
+					return (false, None);
+				}
+
+				self.indices.remove(&weakest_score);
+				self.indices
+					.try_insert(score, insert_idx)
+					.expect("just removed an item, we must be under capacity; qed");
+
+				// ensure that SignedSubmissionsMap never grows past capacity by taking out the
+				// weakest member here.
+				self.map_take(weakest_idx)
+			}
+		};
+
+		// we've taken out the weakest, so update the storage map and the next index
+		self.insertion_overlay.insert(self.next_idx, submission);
+		self.next_idx += 1;
+		(true, weakest)
+	}
+
+	/// Remove the signed submission with the highest score from the set.
+	pub fn pop_last(&mut self) -> Option<SignedSubmissionOf<T>> {
+		let (highest_score, idx) = self.indices.iter().rev().next()?;
+		let (highest_score, idx) = (*highest_score, *idx);
```

Comment:
wonder why https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.pop_last is not stable 🤔